usr/src/uts/common/fs/zfs/vdev.c

```diff
@@ -2489,7 +2489,12 @@ vdev_online(spa_t *spa, uint64_t guid, uint64_t flags, vdev_state_t *newstate)
 		spa_async_request(spa, SPA_ASYNC_CONFIG_UPDATE);
 	}
 
-	if (postevent)
+	if (wasoffline ||
+	    ((oldstate == VDEV_STATE_REMOVED ||
+	    oldstate == VDEV_STATE_CANT_OPEN ||
+	    oldstate == VDEV_STATE_FAULTED) &&
+	    (vd->vdev_state == VDEV_STATE_DEGRADED ||
+	    vd->vdev_state == VDEV_STATE_HEALTHY)))
```

Comment:
Indeed, it looks simpler, but I wanted to show exactly which state transitions we should be notified about. Also, can we be sure that the order of state defines won't change?